### ISSUE TITLE
Scaling all instance weights

### ISSUE THREAD

**Issue by ferraric** [dmlc/xgboost#5915]
XGBoost allows passing a weight for every sample via the DMatrix constructor. I observed that scaling the weight of all samples by a constant factor changed the outcome of the training.

Intuitively I would have guessed that only the relative weight of samples in relation to each other matters.

After some investigation I found that for the min_child_weight parameter the scaled weights imply a different behavior. Also the sample weights make their way into the computation of the gradients and hessian according to this post (https://github.com/dmlc/xgboost/issues/144). There it might also be the case that uniform scaling could lead to a different training outcome.

To me any effect of sample weights other than relative importance is somewhat counterintuitive. Therefore I wanted to ask if this is intended behavior, a bug, or simply unintended consequences? And are there other parameters to control that effect?

If needed, I would be happy to put together a minimal working example to show the effect.

**Comment by thvasilo** [comment]
I'd be interested in a minimal working example.

The way that weights influence the learning also depends on the learning method you choose, see [here](https://github.com/dmlc/xgboost/issues/5822).

> After some investigation I found that for the min_child_weight parameter the scaled weights imply a different behavior.

Could you expand on this a bit?

**Comment by ferraric** [comment]
Thanks for the answer and sorry for the long delay on my part. 

Why I think the min_child_weight parameter has an influence: In the [documentation](https://xgboost.readthedocs.io/en/latest/python/python_api.html#) the parameter is described as: _"Minimum sum of instance weight(hessian) needed in a child"_ I took that to be the instance weight passed to the training. After reading the thread you linked I now believe, as is actually emphasized, that "instance weight" refers to the hessian. Thanks for clearing that confusion already.

As for the minimal example:
```python
import numpy as np
from sklearn.datasets import fetch_openml
from xgboost import XGBClassifier

n_train = 100
X_train = np.random.normal(size=(n_train, 10))
y_train = np.random.randint(low=0, high=9+1, size=n_train)

n_test = 100
X_test = np.random.normal(size=(n_train, 10))
y_test = np.random.randint(low=0, high=9+1, size=n_train)


model = XGBClassifier(tree_method="exact")
weights_one = np.ones(y_train.shape[0])
model.fit(X_train,y_train, sample_weight=weights_one)
predictions_weights_one = model.predict(X_test)

model = XGBClassifier(tree_method="exact")
weights_two = np.ones(y_train.shape[0]) + 1
model.fit(X_train,y_train, sample_weight=weights_two)
predictions_weights_two = model.predict(X_test)

assert (predictions_weights_one == predictions_weights_two).all()
```

Here I would intuitively assume the assertion to pass, since I still weigh all samples equally. But it fails.

**Comment by hcho3** [comment]
That's correct. The `min_child_weight` hyperparameter controls the minimum Hessian value in each node, and every Hessian value gets scaled by the corresponding data weight. This behavior is expected.

**Comment by ferraric** [comment]
Thank you for your answer and confirmation of that aspect. 

Consequently, when I set `min_child_weight` to 0. I would assume that this essentially eliminates any constraints on the minimum Hessian value in each node, correct?

If so, I would then assume that in the example above the assertion should pass if `min_child_weight` is set to 0.
Modified example:

```python
import numpy as np
from xgboost import XGBClassifier

n_train = 100
X_train = np.random.normal(size=(n_train, 10))
y_train = np.random.randint(low=0, high=9+1, size=n_train)

n_test = 100
X_test = np.random.normal(size=(n_train, 10))
y_test = np.random.randint(low=0, high=9+1, size=n_train)


model = XGBClassifier(tree_method="exact", min_child_weight=0)
weights_one = np.ones(y_train.shape[0])
model.fit(X_train,y_train, sample_weight=weights_one)
predictions_weights_one = model.predict(X_test)

model = XGBClassifier(tree_method="exact", min_child_weight=0)
weights_two = np.ones(y_train.shape[0]) + 1
model.fit(X_train,y_train, sample_weight=weights_two)
predictions_weights_two = model.predict(X_test)

assert (predictions_weights_one == predictions_weights_two).all()
```

However, the assertion still fails. This to me indicates that uniformly scaling instance weights has other effects on learning other than via the `min_child_weight` parameter.

**Comment by hcho3** [comment]
Probably. For example, `alpha` (L1 regularization) and `lambda` (L2 regularization) may have an effect. Also, you may run into truncation error due to floating-point arithmetic. So when comparing two floating-point arrays, use
```
np.testing.assert_almost_equal(predictions_weights_one, predictions_weights_two)
```
which will pass when the arrays are sufficiently close to each other in value.

**Comment by ferraric** [comment]
Thanks for the suggestions:

- `alpha` is left as default (0), so IMHO should have no effect
- `lambda` was set to 1, I now set it to 0, this did not change the outcome
- The predictions are integer values (representing discrete classes), so there should not be any rounding issues.
  - When re-running the example 3 times 14%, 12% and 10% of the predictions differed
  - When looking at the raw class outputs on the differing predictions, it is more than rounding errors separating the outputs that lead to the differing predictions

Updated example (for completeness):
```python
import numpy as np
from xgboost import XGBClassifier

n_train = 100
X_train = np.random.normal(size=(n_train, 10))
y_train = np.random.randint(low=0, high=9+1, size=n_train)

n_test = 100
X_test = np.random.normal(size=(n_train, 10))
y_test = np.random.randint(low=0, high=9+1, size=n_train)


model = XGBClassifier(tree_method="exact", min_child_weight=0, reg_lambda=0)
weights_one = np.ones(y_train.shape[0])
model.fit(X_train,y_train, sample_weight=weights_one)
predictions_weights_one = model.predict(X_test)
margins_weights_one = model.predict(X_test, output_margin=True)

model = XGBClassifier(tree_method="exact", min_child_weight=0, reg_lambda=0)
weights_two = np.ones(y_train.shape[0]) + 1
model.fit(X_train,y_train, sample_weight=weights_two)
predictions_weights_two = model.predict(X_test)
margins_weights_two = model.predict(X_test, output_margin=True)

assert (predictions_weights_one == predictions_weights_two).all()
```

I understand that there could still be reasons that uniformly scaling the sample weights leads to different models, as XGBoost is a non-trivial algorithm. However, I maintain that I would expect the models to be the same. Maybe this is a misconception about sample weights on my part, otherwise it would somewhat go against the principle of least surprise.

**Comment by hcho3** [comment]
> I maintain that I would expect the models to be the same. 

I do not think that this expectation is neither reasonable nor feasible. The issue is really the nature of floating-point arithmetic.

The loss term in the gradient boosting algorithm is given by the formula
```
-1/2 * sum_t(G_t / H_t)
```
where `G_t` and `H_t` are the sum of gradients and hessians for the `t`-th leaf node. (I'm leaving out some extra terms for the sake of simplicity.) If the sample weights are uniformly scaled, then both `G_t` and `H_t` terms are scaled uniformly by a constant. What happens is that, in floating-point arithmetic,
```
G/H != (G*alpha) / (H*alpha)
```
The truncation is worse as `alpha` increases in magnitude.

Now the issue is that, if `G/H` term is slightly off, then the gain value for each candidate split (the pair (`feature_id`, `threshold`)) may be slightly off as well. If there are many candidates with similar gains, then the choice of optimal candidate may change.

**Comment by ferraric** [comment]
The statement I made was from the perspective of a user that sees the sample weight parameter as carrying only information about the relative importance of data points.
I understand your reasoning of why this is not feasible. Thanks a lot for your time and insights.

**Comment by hcho3** [comment]
Got it. Let me submit an update to the documentation to document the effect of floating-point arithmetic.

**Comment by HenryBao91** [comment]
I have a question about how to get these  all leaves’  weights from the model ？ I can't find any API to get this. Please help me.
